src/transformers/models/superglue/modeling_superglue.py

```diff
@@ -0,0 +1,796 @@
+        super().__init__()
+        self.descriptor_dim = config.descriptor_dim
+        self.layers_types = config.gnn_layers_types
+        self.layers = nn.ModuleList([SuperGlueAttentionalPropagation(config) for _ in range(len(self.layers_types))])
+
+    def forward(
+        self,
+        descriptors0: torch.Tensor,
+        descriptors1: torch.Tensor,
+        output_attentions: bool = False,
+        output_hidden_states: Optional[bool] = False,
+    ) -> Tuple[torch.Tensor, torch.Tensor, Optional[Tuple], Optional[Tuple]]:
+        all_hidden_states = () if output_hidden_states else None
+        all_attentions = () if output_attentions else None
+
+        if output_hidden_states:
+            new_hidden_state = concat_ragged_pairs((descriptors0,), (descriptors1,))
+            all_hidden_states = all_hidden_states + new_hidden_state
+
+        for gnn_layer, layer_type in zip(self.layers, self.layers_types):
+            if layer_type == "cross":
+                source0, source1 = descriptors1, descriptors0
+            elif layer_type == "self":
+                source0, source1 = descriptors0, descriptors1
+            else:
+                raise ValueError(f"Unrecognized layer type {layer_type}")
+
+            gnn_outputs0 = gnn_layer(
+                descriptors0, source0, output_hidden_states=output_hidden_states, output_attentions=output_attentions
+            )
+            gnn_outputs1 = gnn_layer(
+                descriptors1, source1, output_hidden_states=output_hidden_states, output_attentions=output_attentions
+            )
+
+            delta0 = gnn_outputs0[0]
+            delta1 = gnn_outputs1[0]
+
+            if output_hidden_states:
+                all_hidden_states = all_hidden_states + concat_ragged_pairs(gnn_outputs0[1], gnn_outputs1[1])
+            if output_attentions:
+                all_attentions = all_attentions + concat_ragged_pairs(gnn_outputs0[2], gnn_outputs1[2])
+
+            descriptors0 = descriptors0 + delta0
+            descriptors1 = descriptors1 + delta1
+        return descriptors0, descriptors1, all_hidden_states, all_attentions
+
+
+class SuperGlueFinalProjection(nn.Module):
+    def __init__(self, config: SuperGlueConfig) -> None:
+        super().__init__()
+        descriptor_dim = config.descriptor_dim
+        self.final_proj = nn.Linear(descriptor_dim, descriptor_dim, bias=True)
+
+    def forward(self, descriptors: torch.Tensor) -> torch.Tensor:
+        return self.final_proj(descriptors)
+
+
+class SuperGluePreTrainedModel(PreTrainedModel):
+    """
+    An abstract class to handle weights initialization and a simple interface for downloading and loading pretrained
+    models.
+    """
+
+    config_class = SuperGlueConfig
+    base_model_prefix = "superglue"
+    main_input_name = "pixel_values"
+    supports_gradient_checkpointing = False
+
+    def _init_weights(self, module: Union[nn.Linear, nn.Conv2d, nn.LayerNorm, nn.Conv1d]) -> None:
```

Comment:
Not sure a correct type hint, can be any nn.Module actually, e.g. `SuperGlueMultiLayerPerceptron`